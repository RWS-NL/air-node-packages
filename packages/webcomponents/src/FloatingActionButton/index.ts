export { default as FloatingActionButton, FabProps } from './FloatingActionButton';
