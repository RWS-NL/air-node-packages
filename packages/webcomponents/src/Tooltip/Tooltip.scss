@import '../styles/vars';

.tooltip {
  background-color: $white !important;
  border-radius: 0 !important;
  box-shadow: 0 1px 3px 0 rgb(0 0 0 / 20%), 0 1px 1px 0 rgb(0 0 0 / 14%), 0 2px 1px -1px rgb(0 0 0 / 12%) !important;
  color: rgb(0 0 0 / 87%) !important;
  font-size: 0.8rem !important;
}

.box {
  padding: 8px;
}
