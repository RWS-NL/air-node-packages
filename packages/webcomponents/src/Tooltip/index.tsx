import Box from '@material-ui/core/Box';
import MUITooltip, { TooltipProps as MUITooltipProps } from '@material-ui/core/Tooltip';
import { CSSProperties } from '@material-ui/styles';
import classnames from 'classnames';
import React, { memo } from 'react';
import css from './Tooltip.scss';

export interface TooltipProps extends MUITooltipProps {
  /** Data-qa tag to apply to the tooltip */
  'data-qa'?: string;
  /** Custom CSS classes to pass to the tooltip */
  customclasses?: string | string[];
  /** Any additional CSSProperties to pass to the component */
  style?: CSSProperties;
}

/**
 * Constructs a tooltip using pre-defined Rijkswaterstaat styling
 * @param param Props to pass to the Tooltip
 * @example
 * ```jsx
 * <Tooltip title='Tooltip Content' placement='top'>
 *   <IconButton>
 *     <CloudDownload color='primary' />
 *   </IconButton>
 * </Tooltip>
 * ```
 */
export const Tooltip = memo(
  ({ 'data-qa': qaTag, title, placement, enterDelay, customclasses, style, children, ...props }: TooltipProps) => (
    <MUITooltip
      {...props}
      title={
        <Box data-qa={qaTag} className={css.box}>
          {title}
        </Box>
      }
      placement={placement || 'top'}
      enterDelay={enterDelay || 300}
      classes={{ tooltip: classnames(css.tooltip, customclasses), popper: css.popper }}
      style={style}
    >
      {children}
    </MUITooltip>
  )
);

export default Tooltip;
