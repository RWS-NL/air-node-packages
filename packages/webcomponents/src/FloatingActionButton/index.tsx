import Fab, { FabProps as MFabProps } from '@material-ui/core/Fab';
import { createStyles, makeStyles, Theme } from '@material-ui/core/styles';
import React, { memo, ReactNode } from 'react';
import Tooltip, { TooltipProps } from '../Tooltip';

export interface FabProps extends MFabProps {
  /** The icon to show inside the Floating Action Button */
  icon: ReactNode;
  /** The content to put in the on-hover tooltip */
  tooltipContent: ReactNode;
  /** Props applied to the `Tooltip` element */
  TooltipProps?: Partial<Omit<TooltipProps, 'title'>>;
}

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    fab: {
      position: 'absolute',
      bottom: theme.spacing(2),
      right: theme.spacing(2)
    }
  })
);

/**
 * Constructs a Floating Action Button using pre-defined Rijkswaterstaat styling
 * @param props Props to pass to the Floating Action Button.
 * Also supports all properties applicable to [Material-UI's Fab API](https://material-ui.com/api/fab/)
 * and webcomponents [TooltipProps](https://rws-nl.github.io/air-node-packages/interfaces/_rws_air_webcomponents.tooltipprops.html)
 *
 * @example
 * ```jsx
 * <FloatingActionButton tooltipContent='Cool Tooltip' icon={<SearchIcon/>}/>
 * ```
 */
export const FloatingActionButton = memo(({ tooltipContent, icon, TooltipProps, ...props }: FabProps) => {
  const classes = useStyles();

  return (
    <Tooltip {...TooltipProps} title={tooltipContent}>
      <Fab {...props} className={classes.fab} color='primary'>
        {icon}
      </Fab>
    </Tooltip>
  );
});

export default FloatingActionButton;
