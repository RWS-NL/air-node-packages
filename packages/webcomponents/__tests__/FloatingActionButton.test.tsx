import { Fab } from '@material-ui/core';
import AccessAlarm from '@material-ui/icons/AccessAlarm';
import { FloatingActionButton } from '@src/FloatingActionButton';
import { Tooltip } from '@src/Tooltip';
import { shallow, ShallowWrapper } from 'enzyme';
import React from 'react';
import toJson from "enzyme-to-json";

const mockButtonClick = jest.fn();

describe('Component Tests', () => {
  let floatingActionButton: ShallowWrapper;

  beforeAll(() => {
    floatingActionButton = shallow(
      <FloatingActionButton tooltipContent='Tooltip Sample' onClick={mockButtonClick}>
        <AccessAlarm />
      </FloatingActionButton>
    );
  });

  afterAll(() => {
    jest.clearAllMocks();
  });

  test('GIVEN button click THEN triggers onClick', () => {
    floatingActionButton.find(Fab).simulate('click');
    expect(mockButtonClick).toHaveBeenCalledWith();
    expect(mockButtonClick).toHaveBeenCalledTimes(1);
  });

  test('GIVEN tooltipContent THEN contains content in Tooltip', () => {
    const tooltip = floatingActionButton.find(Tooltip);

    expect(tooltip.prop('title')).toBe('Tooltip Sample');
  });

  test('GIVEN disableToolthen THEN has no Tooltip', () => {
    floatingActionButton = shallow(
      <FloatingActionButton disableTooltip>
        <AccessAlarm />
      </FloatingActionButton>
    );

    const tooltip = floatingActionButton.find(Tooltip);

    expect(tooltip.exists()).toBe(false);
  });
});

describe('Snapshot Testing', () => {
  test('Required Props', () => {
    const floatingActionButton = shallow(
      <FloatingActionButton tooltipContent='Sample' onClick={mockButtonClick}>
        <AccessAlarm />
      </FloatingActionButton>
    );
    expect(toJson(floatingActionButton)).toMatchSnapshot();
  });

  test('Optional Props', () => {
    const floatingActionButton = shallow(
      <FloatingActionButton TooltipProps={{ id: 'tooltip' }} onClick={mockButtonClick}>
        <AccessAlarm />
      </FloatingActionButton>
    );
    expect(toJson(floatingActionButton)).toMatchSnapshot();
  });

  test('Disabling Tooltip', () => {
    const floatingActionButton = shallow(
      <FloatingActionButton disableTooltip>
        <AccessAlarm />
      </FloatingActionButton>
    );

    expect(toJson(floatingActionButton)).toMatchSnapshot();
  });
});
