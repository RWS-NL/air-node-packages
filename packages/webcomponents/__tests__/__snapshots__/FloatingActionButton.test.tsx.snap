// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot Testing Disabling Tooltip 1`] = `
<WithStyles(ForwardRef(Fab))
  className="makeStyles-fab-1"
  color="primary"
>
  <AccessAlarmIcon />
</WithStyles(ForwardRef(Fab))>
`;

exports[`Snapshot Testing Optional Props 1`] = `
<Memo()
  id="tooltip"
  title=""
>
  <WithStyles(ForwardRef(Fab))
    className="makeStyles-fab-1"
    color="primary"
    onClick={[MockFunction]}
  >
    <AccessAlarmIcon />
  </WithStyles(ForwardRef(Fab))>
</Memo()>
`;

exports[`Snapshot Testing Required Props 1`] = `
<Memo()
  title="Sample"
>
  <WithStyles(ForwardRef(Fab))
    className="makeStyles-fab-1"
    color="primary"
    onClick={[MockFunction]}
  >
    <AccessAlarmIcon />
  </WithStyles(ForwardRef(Fab))>
</Memo()>
`;
